### PR TITLE
Main.py cleanup, storage for classification workflow, graphical evaluation for single-label classification

### PR DESCRIPTION
**Pull request by dickeygh**
`Get_Classification.py`:
1. Added elif and an extra else so that the workflow can run for regression and classification with the new `main.py`.

`Analysis.py`:
1. Added `plt.show()` at line 59 to reset the matplotlib plot, so that the single-label classification workflow can plot its graph successfully when also using `impgraph()`.

`Classifiers.py`:
1. Changed 'rf' to 'rfc'

`Features.py`:
1. Added declaration of self.selected_feat_string (line 32) to use in `train.py`.

`Models.py`:
1. Changed multi_label_classification_datasets to be self variable.
2. Added if statement at line 117 so that the workflow in `main.py` could be simplified.

`Storage.py`:
1. Added extra if statement at line 81 so that storage can run properly for multi-label classification.

`Train.py`:
1. Added new function to create roc_curve (lines 177-185)
2. Added lines at 134-141 that create and save a roc graph for single-label classification.

`Main.py`:
1. Major changes were made to `main.py` in this workflow. The changes allow for the entire workflow (classification and regression) to be ran together without prompting the user to select a model type. The workflow has been tested extensively, and both the regression and classification models seem to be running fine. There are several commented blocks of code (lines 23-29, 45-60, 64) that can be used to run specific models/data sets from either the regression or classification workflow.

Overall, this PR adds storage for all classification models and a graphical evaluation for single-label classification models. In addition, the PR simplifies `main.py` by removing if statements and allowing more continuity between the regression and classification workflows.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -18,109 +18,94 @@ def main():
...
                         print()
                         print('Initializing model...', end=' ', flush=True)
                         # initiate model class with algorithm, dataset and target
-                        model1 = models.MlModel(algorithm=alg, dataset=data, target=target, feat_meth=method,
-                                                tune=False, cv=3, opt_iter=25)
+                        opt_iter = 10
+                        if data in ['ESOL.csv', 'Lipophilicity-ID.csv', 'water-energy.csv', 'logP14k.csv',
+                                    'jak2_pic50.csv']:
+                            opt_iter = 25
+                        model = models.MlModel(algorithm=alg, dataset=data, target=target, feat_meth=method,
```
Comment (qle2) [inline main.py:89]: I couldn't quite follow what you were trying to accomplish here. Would you please explain all the pass statements and the different decisions in the `opt_iter` variable?

Comment (dickeygh) [reply main.py:89]: The pass statements are there so that classification models do not attempt to run on regression data sets, and vice versa. I structured the code this way so that both classification and regression could be ran through the same workflow, without if statements differentiating the two. 

The opt_iter variable is decided this way because it was set to 25 for regression and 10 for classification before I made these changes to `main.py`.

File: core/train.py [hunk 2]
```diff
@@ -169,9 +178,43 @@ def train_cls(self, n=5):
     else:
         print('Average accuracy score = %.3f' % stats['acc_avg'], '+- %.3f' % stats['acc_std'])
         print('Average roc_auc score = %.3f' % stats['auc_avg'], '+- %.3f' % stats['auc_std'])
+        print('Average f1_score = %.3f' % stats['f1_score_avg'], '+- %.3f' % stats['f1_score_std'])
         print()
 
+    if self.dataset in ['BBBP.csv', 'bace.csv']:
```
Comment (Dlux804) [inline core/train.py:184]: Again, we can set a variable to achieve this instead of searching a list.  We can have `task_type`  be set to regression, single classification, multiclass, etc. to meet the needs.  Then use this variable instead of searching lists.  These lists are in multiple places in the code and would need to be updated at each instance if we got a new dataset.  We want one single place (master dictionary) to be updated.

File: main.py [hunk 3]
```diff
@@ -19,109 +19,94 @@ def main():
...
+    # sets = {
+    #     'ESOL.csv': 'water-sol',
+    #     'Lipophilicity-ID.csv': 'exp',
+    #     'water-energy.csv': 'expt',
+    #     'logP14k.csv': 'Kow',
+    #     'jak2_pic50.csv': 'pIC50'
+    # }
+
+    for alg in learner: # loop over all learning algorithms
+        feats = Get_Classification.get_classification_feats(alg) # Selects featurizations for classification based on the model being ran, if regression model; uses default featurizations
```
Comment (Dlux804) [inline main.py:64]: I am not sure I understand the point of this.

File: main.py [hunk 4]
```diff
@@ -19,109 +19,94 @@ def main():
...
+                if data in ['BBBP.csv', 'sider.csv', 'clintox.csv', 'bace.csv']:
+                    target = Get_Classification.get_classification_targets(data)
+
+                if data in ['sider.csv', 'clintox.csv'] and alg == 'svc':
+                    pass
+                elif data in ['BBBP.csv', 'sider.csv', 'clintox.csv', 'bace.csv'] and alg in ['ada', 'rf', 'svr', 'gdb', 'nn', 'knn']:
+                    pass
+                elif data in ['ESOL.csv', 'Lipophilicity-ID.csv', 'water-energy.csv', 'logP14k.csv', 'jak2_pic50.csv'] and alg in ['svc', 'knc', 'rfc']:
+                    pass
+                else:
```
Comment (Dlux804) [inline main.py:77]: These if statements based on the dataset can be avoided by using a master dataset dictionary.

```
master_data = {
'ESOL.csv': {task: 'regression', targets: ['water-energy']},
'sider.csv': {task: 'multi_class', targets: ['all', 'the', 'target', 'columns']},
'other.csv': {task: 'single_class', targets: ['column_name']}
}
```

File: main.py (hunk 1 above)
Comment (Dlux804) [reply main.py:90]: There is no technical reason that regression and classification should have different values of `opt_iter` -- it must have been a thing changed during experimenting.

File: core/features.py [hunk 5]
```diff
@@ -30,6 +29,9 @@ def featurize(self, not_silent=True, retrieve_from_mysql=False):
         feat_meth = [int(x) for x in input(
             'Choose your features  by number from list above.  You can choose multiple with \'space\' delimiter:  ').split()]
     selected_feat = [feat_sets[i] for i in feat_meth]
+
+    self.selected_feat_string = '-'.join(selected_feat) # This variable will be used later in train.py for giving classification roc graph a unique file name.
```
Comment (Dlux804) [inline core/features.py:33]: Is this how it is done for regression?

File: core/Get_Classification.py [hunk 6]
```diff
@@ -4,7 +4,7 @@ def get_classification_targets(data):
                    'Eye disorders', 'Investigations', 'Musculoskeletal and connective tissue disorders',
                    'Gastrointestinal disorders',
                    'Social circumstances', 'Immune system disorders', 'Reproductive system and breast disorders',
-                   'Neoplasms benign, malignant and unspecified (incl cysts and polyps)',
+                   'Neoplasms benign, malignant and unspecified',
                    'General disorders and administration site conditions', 'Endocrine disorders',
                    'Surgical and medical procedures', 'Vascular disorders', 'Blood and lymphatic system disorders',
                    'Skin and subcutaneous tissue disorders', 'Congenital, familial and genetic disorders',
```
[the remark this answers is not in the document]
Comment (andreshyer) [reply core/Get_Classification.py:10]: Hey guys, I believe this was changed (or something similar like this) was changed on my PR. MySql is picky about column names, so I had to changed one of them in one of the datasets to work with MySql. I am not sure if this is from that, but I did want to mention

File: core/classifiers.py [hunk 7]
```diff
@@ -10,7 +10,7 @@ def get_classifier(self):
     skl_cls = {
         'svc': SVC,
         'knc': KNeighborsClassifier,
-        'rf': RandomForestClassifier
+        'rfc': RandomForestClassifier
```
[the remark this answers is not in the document]
Comment (andreshyer) [reply core/classifiers.py:13]: This is a super small point, but I this  shows the idea I had before. Instead of using datasets to determine weather or not to use regression or classification, we use the target data column itself. For example, if the target column has continuous values, use regression, and if not use classification. If we were to use this idea, for each algo, we would have a regression and classification pair (rf -> rfr and rfc). 

I am not saying this is the best way to doing things by any means, just an idea.

File: core/storage/storage.py [hunk 8]
```diff
@@ -79,7 +79,7 @@ def store(self):
 
         # grab pandas related objects for export to csv
         if isinstance(v, pd.core.frame.DataFrame) or isinstance(v, pd.core.series.Series):
-            if k == "data":  # Rename target column to a more general one
+            if k == "data" and self.dataset not in self.multi_label_classification_datasets:  # Rename target column to a more general one
```
[the remark this answers is not in the document]
Comment (andreshyer) [reply core/storage/storage.py:82]: This is not looking for a column named data. It is looking for an instance variable named data. Which is our variable that contains all of our featurized data in a dataframe. Then if that dataset only has one target_name column (not in mult classification datasets), then we rename the target column to target. 

I am not sure why this is done, but this was here before I made any PRs I believe.

File: core/analysis.py [hunk 9]
```diff
@@ -56,6 +56,7 @@ def impgraph(self):
     ax.xaxis.grid(False)  # remove just xaxis grid
 
     plt.savefig(self.run_name + '_importance-graph.png')
+    plt.show()
```
[the remark this answers is not in the document]
Comment (andreshyer) [reply core/analysis.py:59]: I second this

File: core/storage/storage.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/storage/storage.py:82]: We want to make sure we don't lose the information about what the target property is.  For example, if we change 'logP' to 'target', someone else might not know what the column represents.  If we don't have a clear reason to rename the column, I suggest we not do it.

File: core/models.py [hunk 10]
```diff
@@ -70,9 +70,8 @@ def __init__(self, algorithm, dataset, target, feat_meth, tune=False, opt_iter=1
         # ingest data.  collect full data frame (self.data)
         # collect pandas series of the SMILES (self.smiles_col)
 
-        if self.dataset in multi_label_classification_datasets:
-            self.data, self.smiles_series = ingest.load_smiles(self, dataset,
-                                                               drop=False)  # Makes drop = False for multi-target classification
+        if self.dataset in self.multi_label_classification_datasets:
+            self.data, self.smiles_series = ingest.load_smiles(self, dataset, drop = False) # Makes drop = False for multi-target classification
```
[the remark this answers is not in the document]
Comment (dickeygh) [reply core/models.py]: I was unable to make this work and did not want to spend too much time on something I couldn't figure out. If you think this is crucial to the workflow, then I can invest more time into making this change.

File: main.py (hunk 4 above)
Comment (dickeygh) [reply main.py]: I was unable to make this work properly, and as a result these if statements are now contained in Get_Task_Type.py. 

Furthermore, Get_Classification.py is still being used to get target columns for classification, since using lists within the dictionary caused issues later in the workflow.

**Comment by dickeygh** [comment]
I just pushed several changes to this PR that address a lot of the issues with the previous iteration. The Get_Task_Type file was added in order to clean up main.py and perform if statements in a different file. The file allows for main.py to skip over algorithm/data set combinations that are not compatible.

I was unable to accomplish this in a different manner, which is why I created this new file.

Furthermore, I was unable to put lists of targets directly into the dictionary for classification, as this caused errors I was unable to solve later in the workflow. As a result, Get_Classification.py is still being used in order to get the target columns for classification data sets.

These changes also included changing self.regressor to self.estimator, moving the graphing functions for classification from train.py to analysis.py, and improving upon the declaration of self.task_type in models.py, as well as using self.task_type instead of searching through lists of datasets in several parts of the workflow.

Making these changes caused the removal of several TODO's from the code.

File: core/analysis.py [hunk 11]
```diff
@@ -16,7 +19,7 @@ def impgraph(self):
     """
 
     # Get numerical feature importances
-    importances2 = self.regressor.feature_importances_  # used later for graph
+    importances2 = self.estimator.feature_importances_  # used later for graph
```
Comment (qle2) [inline core/analysis.py:22]: Good change

File: core/regressors.py [hunk 12]
```diff
@@ -160,7 +160,7 @@ def hyperTune(self, epochs=50,n_jobs=6):
 
     # set up Bayes Search
     bayes = BayesSearchCV(
-        estimator=self.regressor,  # what regressor to use
+        estimator=self.estimator,  # what regressor to use
```
Comment (qle2) [inline core/regressors.py:163]: Good change

**Comment by dickeygh** [comment]
I just pushed several commits to this PR:

`Analysis.py`:
1. I added comments explaining the purpose of the new graphing functions.
2. I changed lines 250-256 to make the graph created have a nicer looking format.

`Get_Task_Type.py`:
1. Added comments explaining the purpose of each if statement, the checker variable, and the function as a whole.
2. Adjusted this file to match 'knc' being changed to 'knn' to match regression workflow.
3. Adjusted this file to match 'svr/svc' being changed to 'svm'.
4. Added extra if statements and task_type variable that is used in main.py to print out the ML task being performed.

`Classifiers.py`:
1. Changed 'knc' to 'knn' to match regression workflow.
2. Changed 'svc' to 'svm' to match new svm workflow.

`Grid.py`:
1. Changed line 259 to be 'svm' to match other changes.

`Name.py`:
1. Added 'svm' to algorithm_list (line 68)

`Regressors.py`:
1. Changed 'svr' to 'svm' to match new workflow.

`Main.py`:
1. Added calling of task_type (line 75) and used at line 84.

File: core/analysis.py [hunk 13]
```diff
@@ -209,3 +213,54 @@ def grid_image(df, filename, molobj=True, smi='smiles'):  # list of molecules to
...
+
+    if self.task_type == 'single_label_classification':
+        # Creates and saves a graphical evaluation for single-label classification
+        fpr, tpr, thresholds = roc_curve(self.test_target, self.predictions)
+        plot_roc_curve(fpr, tpr, self.auc_avg, self.acc_avg, self.f1_score_avg)
+        name = self.algorithm + '-' + self.dataset + '-' + self.selected_feat_string  # Creates a unique file name to save the graph with
+        filename = "%-roc_curve.png" % name
+        plt.legend(loc='best')
+        plt.savefig(self.run_name + '_' + filename)
+        plt.close()
```
Comment (qle2) [inline core/analysis.py:233]: `self.run_name` essentially does the same thing as `name`. Why did you need to add it? Adding `roc_curve.png` already makes it unique

File: core/name.py [hunk 14]
```diff
@@ -65,7 +65,7 @@ def represent_algorithm():
     This representation follows the "Letters and Numbers" rule explained in naming_schemes_v2.pptx in the naming branch
     :return:
     """
-    algorithm_list = ['ada', 'rf', 'svr', 'gdb', 'mlp', 'knn', 'nn', 'svc', 'knc', 'rfc']  # Out current supported algorithm
+    algorithm_list = ['ada', 'rf', 'svr', 'gdb', 'mlp', 'knn', 'nn', 'svc', 'knc', 'rfc', 'svm']  # Out current supported algorithm
```
Comment (qle2) [inline core/name.py:68]: Is `svm` and `svr` the same thing in this case since our pipeline can distinguish between regression and classification?

Comment (dickeygh) [reply core/name.py:68]: Yes, I combined svr and svc into svm, because our pipeline can now distinguish.

File: core/analysis.py (hunk 13 above)
Comment (dickeygh) [reply core/analysis.py:233]: I have it added because I was under the impression it was needed in order for the graph to be saved in the same folder as the other run data. It also helps remove ambiguity about which run the graph is for.

**Comment by dickeygh** [comment]
Just pushed a commit to this PR:


`analysis.py`:
1. Removed name variable, as it is not needed for saving the graphs.

`name.py`:
1. Removed out-dated names from algorithm_list (line 68)
